parfive/downloader.py

```diff
@@ -0,0 +1,436 @@
+            error.
+        """
+        total_files = self.http_queue.qsize() + self.ftp_queue.qsize()
+        done = set()
+        with self._get_main_pb(total_files) as main_pb:
+            if not self.http_queue.empty():
+                done.update(await self._run_http_download(main_pb))
+            if not self.ftp_queue.empty():
+                done.update(await self._run_ftp_download(main_pb))
+
+        # Return one future to represent all the results.
+        return asyncio.gather(*done, return_exceptions=True)
+
+    async def _run_from_queue(self, queue, tokens, main_pb, session=None):
+        futures = []
+        while not queue.empty():
+            get_file = await queue.get()
+            token = await tokens.get()
+            file_pb = self.tqdm if self.file_progress else False
+            future = asyncio.ensure_future(get_file(session, token=token, file_pb=file_pb))
+
+            def callback(token, future, main_pb):
+                tokens.put_nowait(token)
+                # Update the main progressbar
+                if main_pb:
+                    main_pb.update(1)
+
+            future.add_done_callback(partial(callback, token, main_pb=main_pb))
+            futures.append(future)
+
+        return futures
+
+    async def _run_ftp_download(self, main_pb):
+        futures = await self._run_from_queue(self.ftp_queue, self.ftp_tokens, main_pb)
+        # Wait for all the coroutines to finish
+        done, _ = await asyncio.wait(futures)
+
+        return done
+
+    async def _run_http_download(self, main_pb):
+        async with aiohttp.ClientSession(loop=self.loop) as session:
+            futures = await self._run_from_queue(self.http_queue, self.http_tokens,
+                                                 main_pb, session)
+
+            # Wait for all the coroutines to finish
+            done, _ = await asyncio.wait(futures)
+
+            return done
+
+    @staticmethod
+    async def _get_http(session, *, url, filepath_partial, chunksize=100,
+                        file_pb=None, token, overwrite, **kwargs):
+        """
+        Read the file from the given url into the filename given by ``filepath_partial``.
+
+        Parameters
+        ----------
+
+        session : `aiohttp.ClientSession`
+            The `aiohttp.ClientSession` to use to retrieve the files.
+
+        url : `str`
+            The url to retrieve.
+
+        filepath_partial : `callable`
+            A function to call which returns the filepath to save the url to.
+            Takes two arguments ``resp, url``.
+
+        chunksize : `int`
+            The number of bytes to read into the file at a time.
+
+        file_pb : `tqdm.tqdm` or `False`
+            Should progress bars be displayed for each file downloaded.
+
+        token : `parfive.downloader.Token`
+            A token for this download slot.
+
+        kwargs : `dict`
+            Extra keyword arguments are passed to `~aiohttp.ClientSession.get`.
```

Comment:
```suggestion
            Extra keyword arguments are passed to `aiohttp.ClientSession.get`.
```
Doesn't it shrink it down to `.get` otherwise?